tests/unit/s2n_rsa_pss_rsae_test.c

```diff
@@ -241,35 +241,49 @@ int main(int argc, char **argv)
         /* Test: If they share the same RSA key,
          * an RSA cert and an RSA_PSS cert are equivalent for PSS signatures. */
         {
-            struct s2n_pkey rsa_pss_public_key;
-            s2n_pkey_type rsa_pss_pkey_type;
-            EXPECT_OK(s2n_asn1der_to_public_key_and_type(&rsa_pss_public_key, &rsa_pss_pkey_type, &rsa_pss_cert_chain->cert_chain->head->raw));
-            EXPECT_EQUAL(rsa_pss_pkey_type, S2N_PKEY_TYPE_RSA_PSS);
-
-            /* Set the keys equal. */
-            RSA *rsa_key_copy = EVP_PKEY_get1_RSA(rsa_public_key.pkey);
-            POSIX_GUARD_OSSL(EVP_PKEY_set1_RSA(rsa_pss_public_key.pkey, rsa_key_copy), S2N_ERR_KEY_INIT);
+            DEFER_CLEANUP(struct s2n_pkey rsa_public_key_shared = { 0 }, s2n_pkey_free);
+            s2n_pkey_type rsa_public_key_shared_type = S2N_PKEY_TYPE_UNKNOWN;
+            EXPECT_OK(s2n_asn1der_to_public_key_and_type(&rsa_public_key_shared, &rsa_public_key_shared_type,
+                    &rsa_cert_chain->cert_chain->head->raw));
+            EXPECT_EQUAL(rsa_public_key_shared_type, S2N_PKEY_TYPE_RSA);
+
+            DEFER_CLEANUP(struct s2n_pkey rsa_pss_public_key_shared = { 0 }, s2n_pkey_free);
+            s2n_pkey_type rsa_pss_pkey_type_shared = S2N_PKEY_TYPE_UNKNOWN;
+            EXPECT_OK(s2n_asn1der_to_public_key_and_type(&rsa_pss_public_key_shared, &rsa_pss_pkey_type_shared,
+                    &rsa_pss_cert_chain->cert_chain->head->raw));
+            EXPECT_EQUAL(rsa_pss_pkey_type_shared, S2N_PKEY_TYPE_RSA_PSS);
+
+            /* Set the keys equal.
+             *
+             * EVP_PKEY_set1_RSA sets the EVP key type to RSA, even if the set key is RSA_PSS. To
+             * ensure that the RSA_PSS EVP key type remains set to RSA_PSS, the RSA key is
+             * overridden instead of the RSA_PSS key, since its key type is RSA anyway.
+             */
```

Comment:
We talked about how to maybe make this clearer offline:
- different names for "_shared" variables. Maybe as_rsa / as_rsae instead?
- mention in this comment that the key type is checked, and that there is no way to do EVP_PKEY_set1_RSA_PSS